base_layer/wallet/src/ffi.rs

```diff
@@ -0,0 +1,511 @@
+
+// Following methods are to populate the starting state of the wallet from the Clients persistent datastore which MUST
+// be done before starting the Wallet
+
+/// Set the Key Manager
+#[no_mangle]
+pub unsafe extern "C" fn set_key_manager(
+    wallet: *mut WalletFfi,
+    master_key: *const c_char,  // PrivateKey
+    branch_seed: *const c_char, // String,
+    index: c_uint,
+) -> bool
+{
+}
+
+/// Add an output to the wallet. `spent` is a boolean that indicates if this output is a spent or unspent output.
+#[no_mangle]
+pub unsafe extern "C" fn add_output(
+    wallet: *mut WalletFfi,
+    spent: char,                 // Bool,
+    value: c_ulonglong,          // u64
+    spending_key: *const c_char, // PrivateKey,
+    feature_flags: c_char,       // OutputFlags,
+    maturity: c_ulonglong,       // u64
+) -> bool
+{
+}
+
+/// Initialize a PendingTransactionOutputs struct to be populated
+#[no_mangle]
+pub unsafe extern "C" fn create_pending_transaction_outputs(
+    tx_id: c_ulonglong,       // u64
+    timestamp: *const c_char, // NaiveDateTime
+) -> *mut PendingTransactionOutputs
+{
+    Box::into_raw(Box::new(PendingTransactionOutputs {
+        tx_id,
+        outputs_to_be_spent: Vec::new(),
+        outputs_to_be_received: Vec::new(),
+        timestamp: NaiveDateTime::parse_from_str("timestamp", "THE FORMAT WE CHOOSE").unwrap_or(NaiveDateTime::now), /* TODO Select a datetime format */
+    }))
+}
+
+/// Append an UnblindedOutput to be spent to the pending transaction outputs object
+#[no_mangle]
+pub unsafe extern "C" fn add_output_to_spend(
+    pending_tx: *mut PendingTransactionOutputs,
+    value: c_ulonglong,
+    spending_key: *const c_char,
+    feature_flags: c_char, // OutputFlags,
+    maturity: c_ulonglong, // u64
+) -> bool
+{
+    // append this output to PendingTransactionOutputs.outputs_to_be_spent
+}
+
+/// Append an UnblindedOutput to be recieved to the pending transaction outputs object
+#[no_mangle]
+pub unsafe extern "C" fn add_output_to_received(
+    pending_tx: *mut PendingTransactionOutputs,
+    value: c_ulonglong,
+    spending_key: *const c_char,
+    feature_flags: c_char, // OutputFlags,
+    maturity: c_ulonglong, // u64
+) -> bool
+{
+    // append this output to PendingTransactionOutputs.outputs_to_be_received
+}
+
+/// Add an output to the wallet. `spent` is a boolean that indicates if this output is a spent or unspent output.
+#[no_mangle]
+pub unsafe extern "C" fn add_pending_transaction_outputs(
+    wallet: *mut WalletFfi,
+    pending_tx: *mut PendingTransactionOutputs,
+) -> bool
+{
+    // append this data to the wallet
+}
+
+/// Initialize a Transaction struct to be populated
+#[no_mangle]
+pub unsafe extern "C" fn create_transaction(
+    tx_id: c_ulonglong,    // u64
+    offset: *const c_char, // PrivateKey
+) -> *mut Transaction
+{
+}
+
+/// Add a transaction input to a transaction struct
+#[no_mangle]
+pub unsafe extern "C" fn add_transaction_input(
```

Comment:
I'm sure you've thought about this, but are these all strictly necessary?

My thinking is that the client doesn't micromanage the TX creation process like this. 

Client: send_tari([external_amounts], [my_amounts], lock_height, fee_per_gram etc)
libwallet: creates new spend keys, kicks off Tx protocol, signs, sends tx to recipient and informs client about new utxos
Client: stores UTXOs in persistence

Receiver: Gets tx (possibly ask client for perimission to continue?); creates new output, signs and responds. Needs some way to inform client (Is there some way for client to establish a long-held callback function for this?)

Sender libwallet: Receives tx; sings and finalizes it (possibly ask client for confirmation?); broadcast tx, and inform client about result via callback.

So throughout this process, as much as possible is kept inside the libwallet black box, and so I'm not sure we need `add_transaction_input` etc. At least in the MVP case.

Reply:
I see that all this is exposed below. I'm not sure why these functions need to be exposed in the ffi.

Reply:
This is all due to the fact that the persistence live in the Client side of the FFI so these functions are there for the Client to initially populate the memory models of LibWallet. In other words this is how, on start up, the FFI client reads the state from their Datastore and then pushes it into the LibWallet for the LibWallet to do its thing.